node/pkg/p2p/run_params.go

```diff
@@ -0,0 +1,220 @@
+	gst *common.GuardianSetState,
+	rootCtxCancel context.CancelFunc,
+	opts ...RunOpt,
+) (*RunParams, error) {
+	p := &RunParams{
+		bootstrapPeers: bootstrapPeers,
+		networkID:      networkID,
+		priv:           priv,
+		gst:            gst,
+		rootCtxCancel:  rootCtxCancel,
+	}
+
+	for _, opt := range opts {
+		err := opt(p)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := p.verify(); err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// WithSignedObservationListener is used to set the channel to receive `SignedObservation“ messages.
+func WithSignedObservationListener(obsvC chan<- *common.MsgWithTimeStamp[gossipv1.SignedObservation]) RunOpt {
+	return func(p *RunParams) error {
+		p.obsvC = obsvC
+		return nil
+	}
+}
+
+// WithSignedVAAListener is used to set the channel to receive `SignedVAAWithQuorum messages.
+func WithSignedVAAListener(signedInC chan<- *gossipv1.SignedVAAWithQuorum) RunOpt {
+	return func(p *RunParams) error {
+		p.signedInC = signedInC
+		return nil
+	}
+}
+
+// WithObservationRequestListener is used to set the channel to receive `ObservationRequest messages.
+func WithObservationRequestListener(obsvReqC chan<- *gossipv1.ObservationRequest) RunOpt {
+	return func(p *RunParams) error {
+		p.obsvReqC = obsvReqC
+		return nil
+	}
+}
+
+// WithChainGovernorConfigListener is used to set the channel to receive `SignedChainGovernorConfig messages.
+func WithChainGovernorConfigListener(signedGovCfg chan *gossipv1.SignedChainGovernorConfig) RunOpt {
+	return func(p *RunParams) error {
+		p.signedGovCfg = signedGovCfg
+		return nil
+	}
+}
+
+// WithChainGovernorStatusListener is used to set the channel to receive `SignedChainGovernorStatus messages.
+func WithChainGovernorStatusListener(signedGovSt chan *gossipv1.SignedChainGovernorStatus) RunOpt {
+	return func(p *RunParams) error {
+		p.signedGovSt = signedGovSt
+		return nil
+	}
+}
+
+// WithDisableHeartbeatVerify is used to set disableHeartbeatVerify.
+func WithDisableHeartbeatVerify(disableHeartbeatVerify bool) RunOpt {
+	return func(p *RunParams) error {
+		p.disableHeartbeatVerify = disableHeartbeatVerify
+		return nil
+	}
+}
+
+// WithGuardianOptions is used to set options that are only meaningful to the guardian.
+func WithGuardianOptions(
+	nodeName string,
+	gk *ecdsa.PrivateKey,
+	obsvC chan<- *common.MsgWithTimeStamp[gossipv1.SignedObservation],
+	signedInC chan<- *gossipv1.SignedVAAWithQuorum,
+	obsvReqC chan<- *gossipv1.ObservationRequest,
+	gossipSendC chan []byte,
+	obsvReqSendC <-chan *gossipv1.ObservationRequest,
+	acct *accountant.Accountant,
+	gov *governor.ChainGovernor,
+	disableHeartbeatVerify bool,
+	components *Components,
+	ibcFeaturesFunc func() string,
+	gatewayRelayerEnabled bool,
+	ccqEnabled bool,
+	signedQueryReqC chan<- *gossipv1.SignedQueryRequest,
+	queryResponseReadC <-chan *query.QueryResponsePublication,
+	ccqBootstrapPeers string,
+	ccqPort uint,
+	ccqAllowedPeers string,
+) RunOpt {
+	return func(p *RunParams) error {
```

Comment:
Do you have to return an error type for this?  It's always nil.

Reply:
That is the signature of `RunOpt`. Although none of those return an error at this point, some might some day, so I would like to leave it. Also, that is the pattern that I copied from libp2p...